File: Medium/2001-jump-game-vii/jump-game-vii.py
class Solution:
    def solve(self, idx, s, mn, mx, dp, n):
        if idx>=n or s[idx] == '1':
            return False

        if idx == n-1:
            return True

        if dp[idx] != -1:
            return dp[idx]

        dp[idx]=False
        for i in range(mn, mx+1):
            if self.solve(idx+i, s, mn, mx, dp, n):
                dp[idx]=True
                return True
    
        return dp[idx]

    def canReach(self, s: str, mn: int, mx: int) -> bool:
        n=len(s)
        if s[0] == '1' or s[n-1] == '1':
            return False
        c=0
        for ch in s:
            if ch == '1':
                c+=1
                if c>=mx:
                    return False
            else:
                c=0

        dp=[-1]*n
        return self.solve(0, s, mn, mx, dp, n)


        

File: Medium/2001-jump-game-vii/jump-game-vii.cpp
class Solution{
public:
    bool solve(int idx, string &s, int mn, int mx, vector<int> &dp, int n){
        if(idx>=n || s[idx] == '1')
            return false;

        if(idx == n-1)
            return true;

        if(dp[idx] != -1)
            return dp[idx];

        dp[idx]=false;
        for(int i=mn; i<=mx; i++){
            if(solve(idx+i, s, mn, mx, dp, n)){
                return dp[idx]=true;
            }
        }

        return dp[idx];
    }

    bool canReach(string s, int mn, int mx){
        int n=s.size();
        if(s[0] == '1' || s[n-1] == '1')
            return false;
        int c=0;
        for(int i=0; i<n; i++){
            if(s[i] == '1'){
                c++;
                if(c>=mx)
                    return false;
            }
            else{
                c=0;
            }
        }

        vector<int> dp(n, -1);
        return solve(0, s, mn, mx, dp, n);


    }
};